File: Algorithms/p230_Kth_Smallest_Element_in_a_BST_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        return kthSmallestStep(root,k);
    }
    int kthSmallestStep(TreeNode* node, int& k){
        if(node){
            int val = kthSmallestStep(node->left,k);
            if(k==0){
                return val;
            }
            if(--k==0){
                return node->val;
            }
            return kthSmallestStep(node->right,k);
        }else{
            return -1; // dummy return value
        }
    }
};

File: Algorithms/p230_Kth_Smallest_Element_in_a_BST/p230_Kth_Smallest_Element_in_a_BST_1.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        return self.kthSmallestStep(root,[k])
        
    def kthSmallestStep(self, node, k_l):
        if node:
            val = self.kthSmallestStep(node.left,k_l)
            if k_l[0]==0:
                return val
            k_l[0] -= 1
            if k_l[0]==0:
                return node.val
            return self.kthSmallestStep(node.right,k_l)
        else:
            return -1 # dummy return value
